libV2/vault.js

```diff
@@ -0,0 +1,144 @@
+     * @return {undefined}
+    */
+    authenticateV4Request(params, requestContexts, callback) {
+        const {
+            accessKey, signatureFromRequest, region, scopeDate,
+            stringToSign,
+        } = params.data;
+        const { log } = params;
+        log.debug('authenticating V4 request');
+        const serializedRCs = requestContexts.map(rc => rc.serialize());
+        this._client.verifySignatureV4(
+            stringToSign, signatureFromRequest,
+            accessKey, region, scopeDate,
+            { reqUid: log.getSerializedUids(), requestContext: serializedRCs },
+            (err, authInfo) => {
+                if (err) {
+                    log.trace('error from vault', { error: err });
+                    return callback(err);
+                }
+                return callback(null,
+                    authInfo.message.body.authorizationResults);
+            },
+        );
+    }
+
+    /**
+    * Returns canonical Ids for a given list of account Ids
+    * @param {string[]} accountIds - list of account ids
+    * @param {object} log - Werelogs request logger
+    * @param {callback} callback - callback with error and result as params
+    * @return {undefined}
+    */
+    getCanonicalIds(accountIds, log, callback) {
+        log.debug('retrieving canonical ids for account ids', {
+            method: 'Vault.getCanonicalIds',
+        });
+        return this._client.getCanonicalIdsByAccountIds(accountIds,
+            { reqUid: log.getSerializedUids(), logger: log }, callback);
+    }
+}
+
+const handler = new Vault(config);
+auth.setHandler(handler);
+
+async function authenticateRequest(request, params) {
+
+    const policyContext = new policies.RequestContext(
+        request.headers,
+        params.resource,
+        params.body[params.resource],
+        request.ip,
+        request.ctx.encrypted,
+        params.Action.value,
+        'utapi',
+    );
+
+    return new Promise((resolve, reject) => {
+        auth.server.doAuth(request, request.logger, (err, res) => {
+            if (err && (err.InvalidAccessKeyId || err.AccessDenied)) {
+                resolve([false]);
+                return;
+            }
+            if (err) {
+                reject(err);
+                return;
+            }
+            // Will only have res if request is from a user rather than an account
+            if (res) {
+                try {
+                    const authorizedResources = (res || [])
+                        .reduce((authed, result) => {
+                            if (result.isAllowed) {
+                                // result.arn should be of format:
+                                // arn:scality:utapi:::resourcetype/resource
+                                assert(typeof result.arn === 'string');
+                                assert(result.arn.indexOf('/') > -1);
+                            }
+                            const resource = result.arn.split('/')[1];
```

Comment:
Does this always work in case `! result.isAllowed`?

Reply:
whoops, those should be inside the conditional